addons/pos_stripe/models/pos_payment_method.py

```diff
@@ -0,0 +1,110 @@
+import logging
+import requests
+
+from odoo import fields, models, api, _
+from odoo.exceptions import ValidationError, UserError
+from requests.exceptions import ConnectTimeout
+
+_logger = logging.getLogger(__name__)
+TIMEOUT = 10
+
+class PosPaymentMethod(models.Model):
+    _inherit = 'pos.payment.method'
+
+    def _get_payment_terminal_selection(self):
+        return super()._get_payment_terminal_selection() + [('stripe', 'Stripe')]
+
+    # Stripe
+    stripe_serial_number = fields.Char(help='[Serial number of the stripe terminal], for example: WSC513105011295', copy=False)
+
+    @api.constrains('stripe_serial_number')
+    def _check_stripe_serial_number(self):
+        if not self._get_stripe_secret_key():
+            raise ValidationError(_('Complete the Stripe onboarding.'))
+
+        for payment_method in self:
+            if not payment_method.stripe_serial_number:
+                continue
+            existing_payment_method = self.search([('id', '!=', payment_method.id),
+                                                   ('stripe_serial_number', '=', payment_method.stripe_serial_number)],
+                                                  limit=1)
+            if existing_payment_method:
+                raise ValidationError(_('Terminal %s is already used on payment method %s.')
+                                      % (payment_method.stripe_serial_number, existing_payment_method.display_name))
+
+    @api.model
+    def _get_stripe_secret_key(self):
+        stripe_secret_key = self.env['payment.acquirer'].search([('provider', '=', 'stripe')], limit=1).stripe_secret_key
+
+        if not stripe_secret_key:
+            raise ValidationError(_('Complete the Stripe onboarding.'))
+
+        return stripe_secret_key
+
+    @api.model
+    def stripe_connection_token(self):
+        endpoint = 'https://api.stripe.com/v1/terminal/connection_tokens'
+
+        try:
+            req = requests.post(endpoint, auth=(self._get_stripe_secret_key(), ''), timeout=TIMEOUT)
+        except ConnectionError:
+            raise UserError(_("Connection lost between Odoo and Stripe."))
+        except ConnectTimeout:
+            raise UserError(_("There are some connection issues between us and Stripe, try again later."))
+
+        if req.status_code == 401:
+            return {
+                'error': {
+                    'status_code': req.status_code,
+                    'message': req.text
+                }
+            }
+
+        return req.json()
+
+    @api.model
+    def stripe_payment_intent(self, currency, amount):
+        # For Terminal payments, the 'payment_method_types' parameter must include
+        # 'card_present' and the 'capture_method' must be set to 'manual'
+        endpoint = 'https://api.stripe.com/v1/payment_intents'
+
+        try:
+            data = ("currency=%s&amount=%s&payment_method_types[]=card_present&capture_method=manual") % (currency, int(amount * 100))
```

Comment:
Is this safe considering `currency` is under complete caller control? They can inject anything they want in there. Same issue with `paymentIntentId` below.

Reply:
For amount, int(amount/currency.rounding)

Reply:
I don't understand this answer. `currency` is formatted as-is into the request body with no escaping, I've no idea how stripe would react to e.g. large bodies, or finding a `\0`, or a bunch of newlines, ...  so assuming a worst-case scenario whoever calls the method might have complete control over the body except for the `currency=` prefix.